src/Datadog.Trace/Agent/AgentWriter.cs

```diff
@@ -62,57 +118,295 @@ public async Task FlushAndCloseAsync()
                 return;
             }
 
-            await Task.WhenAny(_flushTask, Task.Delay(TimeSpan.FromSeconds(20)))
-                      .ConfigureAwait(false);
+            _serializationMutex.Set();
+
+            var delay = Task.Delay(TimeSpan.FromSeconds(20));
+
+            var completedTask = await Task.WhenAny(_serializationTask, delay)
+                .ConfigureAwait(false);
 
-            if (!_flushTask.IsCompleted)
+            if (completedTask != delay)
             {
-                Log.Warning("Could not flush all traces before process exit");
+                await Task.WhenAny(_flushTask, Task.Delay(TimeSpan.FromSeconds(20)))
+                    .ConfigureAwait(false);
+
+                if (_frontBuffer.TraceCount == 0 && _backBuffer.TraceCount == 0)
+                {
+                    // All good
+                    return;
+                }
+
+                // In some situations, the flush thread can exit before flushing all the threads
+                // Force a flush for the leftover traces
+                completedTask = await Task.WhenAny(Task.Run(() => FlushBuffers(flushAllBuffers: true)), delay)
```

Comment:
For the `delay` variable here, did you mean to re-use the original `Task.Delay` or did you intend to re-assign a new `Task.Delay`?

Reply:
I meant to re-use the original delay. The timeout is for the whole `FlushAndCloseAsync` operation